### DISCUSSION TITLE
About SQL editor

### DISCUSSION THREAD

**Discussion by m0ast** [supabase/supabase#1571] [Questions] [answered]
Hello,
I tried to have autocompletion for SQL in vscode but got unsatisfying results, much less useful than Supabase Editor. Though it might not help I was wondering if you would be inclined to share some of the technical implementations you used for Monaco.
Btw many thanks for supabase, I really like it.

**Comment by quicksnap** [comment]
DBeaver has OK autocomplete: https://dbeaver.io/

It's UI is very Java clunky, and if you have Apple Silicon, it's not native yet. But it's very powerful!

**Comment by soedirgo** [comment, marked as answer]
@m0ast have you tried https://github.com/Borvik/vscode-postgres? We just ported it for the SQL editor IIRC.

**Reply by m0ast** [reply]
Yeah, this may be the good one, thank you! Perhaps you know why there is some difference with inferred completions and tooltip icons?

**Reply by soedirgo** [reply]
Hmm, I'm not familiar with how it's implemented - perhaps @phamhieu can chime in?

**Reply by phamhieu** [reply]
> Yeah, this may be the good one, thank you! Perhaps you know why there is some difference with inferred completions and tooltip icons?

it's depended on the kind of completion: https://microsoft.github.io/monaco-editor/api/enums/monaco.languages.completionitemkind.html

**Comment by phamhieu** [comment]
> Though it might not help I was wondering if you would be inclined to share some of the technical implementations you used for Monaco

The implementation is simple. 
- Pulling data related to : schema, tableColumns, functions and SQL keyword from db
- implement code completion: https://github.com/Borvik/vscode-postgres/blob/1e6cc18e0bece575a96594c248680e625396a63a/src/language/server.ts#L427
- implement signature help for function: https://github.com/Borvik/vscode-postgres/blob/1e6cc18e0bece575a96594c248680e625396a63a/src/language/server.ts#L594
